fix(hc): Use RpcOrganization in Pipeline

RyanSkonnord opened the pull request:
This is a step toward allowing views that depend on the auth pipeline to run in the control silo mode.

Add an HC service method for updating Organization flags, replacing a write operation done by `AuthHelper.disable_2fa_required`.

RyanSkonnord committed:
fix(hc): Use RpcOrganization in Pipeline

This is a step toward allowing views that depend on the auth pipeline to
run in the control silo mode.

Add an HC service method for updating Organization flags, replacing a
write operation done by `AuthHelper.disable_2fa_required`.

RyanSkonnord committed:
Tweak import style

markstory commented on src/sentry/services/hybrid_cloud/organization/impl.py:
```diff
@@ -281,6 +282,18 @@ def _query_organizations(
+            elif value is False:
+                updates = updates.bitand(~Organization.flags[name])
+            else:
+                raise TypeError
```
```suggestion
                raise TypeError(f"Invalid value {value} received for update_flags")
```

Might be helpful for future us.

markstory reviewed (approved):
Looks good to me, but the test failures seem relevant.

RyanSkonnord committed:
Better TypeError message

RyanSkonnord committed:
Set Migrator param to RpcOrganization

RyanSkonnord committed:
Adapting integration pipeline and providers to RpcOrganization

RyanSkonnord committed:
Change post_install signature

RyanSkonnord committed:
Scaffolding for RepositoryService

RyanSkonnord committed:
Replace Repository filter queries with HC service

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Fix MigratorTest setup

RyanSkonnord committed:
Introduce ProjectService.set_option

RyanSkonnord committed:
Fix Organization query in migrate_repo

RyanSkonnord committed:
Use OrganizationService in AuthOrganizationLoginView

RyanSkonnord committed:
Add ProjectService.get_option

RyanSkonnord committed:
Fix import style

Co-authored-by: Mark Story <mark@mark-story.com>

getsantry[bot] committed:
style(lint): Auto commit lint changes

RyanSkonnord committed:
Fix IntegrationTestCase setup

RyanSkonnord committed:
Fix some auth-related silo problems

RyanSkonnord committed:
Fix query in OrganizationAuthSettingsView

RyanSkonnord committed:
Extract OrganizationAbsoluteUrlMixin

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Add src/sentry/types/organization.py to mypy.ini

markstory commented on src/sentry/services/hybrid_cloud/project/impl.py:
```diff
@@ -0,0 +1,17 @@
+
+
+class DatabaseBackedProjectService(ProjectService):
+    def get_option(self, *, project: RpcProject, key: str) -> Union[str, int, bool]:
+        orm_project = Project.objects.get(id=project.id)
+        result = ProjectOption.objects.get_value(orm_project, key, None)
+        return result  # type: ignore
```
> I think the biggest drawback would be that we could no longer have a simple function with the signature

That's a good point. Having that interface stay simple is more important than optimizing away a few option reads that cross silos.

markstory commented on src/sentry/tasks/integrations/migrate_repo.py:
```diff
@@ -47,5 +48,5 @@ def migrate_repo(repo_id: int, integration_id: int, organization_id: int) -> Non
 
         Migrator.run(
-            integration=integration, organization=Organization.objects.get(id=organization_id)
+            integration=integration, organization=organization_service.get(id=organization_id)
```
Does the `Repository` query on line 16 of this file also need to be updated? It looks like this task will run on control silo, as it is spawned from the integration's `post_install` hook method.

markstory commented on src/sentry/integrations/bitbucket_server/integration.py:
```diff
@@ -305,17 +310,22 @@ class BitbucketServerIntegrationProvider(IntegrationProvider):
             organization_id=organization.id,
-            provider__in=["bitbucket_server", "integrations:bitbucket_server"],
-            integration_id__isnull=True,
-        ).values_list("id", flat=True)
+            providers=["bitbucket_server", "integrations:bitbucket_server"],
+            has_integration=False,
+        )
 
-        for repo_id in repo_ids:
+        for repo in repos:
             migrate_repo.apply_async(
```
We might want to make this into an RPC/outbox task. The `migrate_repo` task does a bunch of write operations to Repository and plugin settings.

markstory commented on src/sentry/integrations/vsts/integration.py:
```diff
@@ -356,19 +359,19 @@ class VstsIntegrationProvider(IntegrationProvider):  # type: ignore
+        )
 
-        for repo_id in repo_ids:
+        for repo in repos:
             migrate_repo.apply_async(
```
We might want to make this into an RPC/outbox message as well, similar to other repo migrations.

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Accept Organization as Pipeline constructor arg; serialize on init

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Introduce ProjectService

RyanSkonnord committed:
Introduce OptionMixin

RyanSkonnord committed:
Fix some import errors

RyanSkonnord committed:
Kludge metaclass conflict by removing ABC inheritance

RyanSkonnord committed:
Implement RpcProject option methods

RyanSkonnord committed:
Implement RpcOrganization option methods

RyanSkonnord committed:
Make OrganizationOptionManager.get_value consistent with interface

This causes it to accept the validation function as an unused argument.
Could future-proof it by applying it the same way ProjectOption does?

RyanSkonnord committed:
Merge branch 'master' into model-option-interface

RyanSkonnord committed:
Preserve bool return type from update_option

RyanSkonnord committed:
Merge branch 'master' into model-option-interface

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Merge branch 'model-option-interface' into rpc-organization-in-pipeline

RyanSkonnord committed:
Grr names

RyanSkonnord committed:
Introduce update_default_role to replace organization.save()

RyanSkonnord committed:
Implement multi-get on RPC services

RyanSkonnord committed:
Merge branch 'master' into model-option-interface

RyanSkonnord committed:
Roll back unneeded multi-get methods

Per discussion on Slack. Keep some type fixes from the previous commit.

RyanSkonnord committed:
Missed a return sig

RyanSkonnord committed:
get_from_cache (suggestions from code review)

Co-authored-by: Mark Story <mark@mark-story.com>

RyanSkonnord committed:
Merge branch 'master' into model-option-interface

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Fix queries in organization_auth_settings

RyanSkonnord committed:
Merge branch 'master' into model-option-interface

RyanSkonnord committed:
Merge branch 'model-option-interface' into rpc-organization-in-pipeline

RyanSkonnord committed:
Fix query error in AwsLambdaProjectSelectPipelineView

RyanSkonnord committed:
Fix AwsLambdaIntegrationTest failures

RyanSkonnord committed:
Merge branch 'master' into rpc-organization-in-pipeline

RyanSkonnord committed:
Have RpcRepository match nullable external_id

RyanSkonnord committed:
Fix repository_service call in CustomSCMIntegration

RyanSkonnord committed:
Fix assertion in test_renders_basic

dashed committed:
Merge branch 'master' into rpc-organization-in-pipeline

dashed committed:
check org status again since org can be cached

dashed committed:
wtf?

dashed committed:
fix bug
